framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -1007,7 +1007,10 @@ protected void checkPurity(MethodTree node) {
     if (suggestPureMethods && !TreeUtils.isSynthetic(node)) {
       // Issue a warning if the method is pure, but not annotated as such.
       EnumSet<Pure.Kind> additionalKinds = r.getKinds().clone();
-      additionalKinds.removeAll(kinds);
+      if (!infer) {
+        // WPI output should contain existing purity annotations
+        additionalKinds.removeAll(kinds);
+      }
```

Comment:
@mernst apparently the reason I was able to figure the issue this morning out so quickly because this change in the all-systems PR is almost identical. This hunk is why the tests for this PR were passing, but master was failing. (Sorry I forgot that I'd made this change in this PR - that's my fault for not realizing that this hunk needed to be merged earlier with the changes to DLJC.)